File: 188-best-time-to-buy-and-sell-stock-iv/best-time-to-buy-and-sell-stock-iv.java
class Solution {
    public int maxProfit(int k, int[] prices) {
        if (k == 0) return 0;
        int[] costs = new int[k + 1];
        int[] profits = new int[k + 1];
        Arrays.fill(costs, Integer.MAX_VALUE);

        for (int price : prices) {
            for (int i = 1; i <= k; i++) {
                costs[i] = Math.min(costs[i], price - profits[i - 1]);
                profits[i] = Math.max(profits[i], price - costs[i]);
            }
        }
        return profits[k];
    }
}

File: 188-best-time-to-buy-and-sell-stock-iv/best-time-to-buy-and-sell-stock-iv.py
class Solution:
    def maxProfit(self, k: int, prices: List[int]) -> int:
        if k == 0:
            return 0

        profit = [0] * (k + 1)
        cost = [float('inf')] * (k + 1)

        for price in prices:
            for i in range(k):
                cost[i + 1] = min(cost[i + 1], price - profit[i])
                profit[i + 1] = max(profit[i + 1], price - cost[i + 1])
        return profit[k]